File: api/crs/codegen/lib/crs_prodAtAx/crs_prodAtAx.c
#include "crs_prodAtAx.h"
#include "crs_prodAtAx_types.h"
#include "m2c.h"
#include "omp.h"
#include <math.h>

static void b_crs_prodAtAx(const emxArray_int32_T *A_row_ptr,
                           const emxArray_int32_T *A_col_ind,
                           const emxArray_real_T *A_val, int A_nrows,
                           int A_ncols, const emxArray_real_T *x,
                           emxArray_real_T *b);

static void b_crs_prodAtx(const emxArray_int32_T *A_row_ptr,
                          const emxArray_int32_T *A_col_ind,
                          const emxArray_real_T *A_val, int A_nrows,
                          int A_ncols, const emxArray_real_T *x,
                          emxArray_real_T *b);

static void b_crs_prodAtx_kernel(const emxArray_int32_T *row_ptr,
                                 const emxArray_int32_T *col_ind,
                                 const emxArray_real_T *val,
                                 const emxArray_real_T *x, int x_m,
                                 emxArray_real_T *b, int b_m, int nrows,
                                 int ncols, int nrhs, boolean_T varargin_1);

static void b_crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                                const emxArray_int32_T *col_ind,
                                const emxArray_real_T *val,
                                const emxArray_real_T *x, int x_m,
                                emxArray_real_T *b, int b_m, int nrows,
                                int nrhs, boolean_T varargin_1);

static void b_m2c_error(void);

static void c_crs_prodAtAx(const emxArray_int32_T *A_row_ptr,
                           const emxArray_int32_T *A_col_ind,
                           const emxArray_real_T *A_val, int A_nrows,
                           int A_ncols, const emxArray_real_T *x,
                           emxArray_real_T *b);

static void c_m2c_error(void);

static void crs_prodAtx(const emxArray_int32_T *A_row_ptr,
                        const emxArray_int32_T *A_col_ind,
                        const emxArray_real_T *A_val, int A_nrows, int A_ncols,
                        const emxArray_real_T *x, emxArray_real_T *b);

static void crs_prodAtx_kernel(const emxArray_int32_T *row_ptr,
                               const emxArray_int32_T *col_ind,
                               const emxArray_real_T *val,
                               const emxArray_real_T *x, int x_m,
                               emxArray_real_T *b, int b_m, int nrows,
                               int ncols, int nrhs);

static void crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                              const emxArray_int32_T *col_ind,
                              const emxArray_real_T *val,
                              const emxArray_real_T *x, int x_m,
                              emxArray_real_T *b, int b_m, int nrows, int nrhs);

static void d_m2c_error(void);

static void m2c_error(void);

static void m2c_warn(void);

static void b_crs_prodAtAx(const emxArray_int32_T *A_row_ptr,
                           const emxArray_int32_T *A_col_ind,
                           const emxArray_real_T *A_val, int A_nrows,
                           int A_ncols, const emxArray_real_T *x,
                           emxArray_real_T *b)
{
  emxArray_real_T *Ax;
  int i;
  int loop_ub;
  emxInit_real_T(&Ax, 2);
  i = Ax->size[0] * Ax->size[1];
  Ax->size[0] = A_nrows;
  Ax->size[1] = x->size[1];
  emxEnsureCapacity_real_T(Ax, i);
  loop_ub = A_nrows * x->size[1];
  for (i = 0; i < loop_ub; i++) {
    Ax->data[i] = 0.0;
  }
  crs_prodAx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], Ax, A_nrows,
                    A_nrows, x->size[1]);
  i = b->size[0] * b->size[1];
  b->size[0] = A_ncols;
  b->size[1] = x->size[1];
  emxEnsureCapacity_real_T(b, i);
  loop_ub = A_ncols * x->size[1];
  for (i = 0; i < loop_ub; i++) {
    b->data[i] = 0.0;
  }
  b_crs_prodAtx(A_row_ptr, A_col_ind, A_val, A_nrows, A_ncols, Ax, b);
  emxFree_real_T(&Ax);
}

static void b_crs_prodAtx(const emxArray_int32_T *A_row_ptr,
                          const emxArray_int32_T *A_col_ind,
                          const emxArray_real_T *A_val, int A_nrows,
                          int A_ncols, const emxArray_real_T *x,
                          emxArray_real_T *b)
{
  if ((b->size[0] < A_ncols) || (b->size[1] < x->size[1])) {
    d_m2c_error();
  }
  crs_prodAtx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], b, b->size[0],
                     A_nrows, A_ncols, x->size[1]);
}

static void b_crs_prodAtx_kernel(const emxArray_int32_T *row_ptr,
                                 const emxArray_int32_T *col_ind,
                                 const emxArray_real_T *val,
                                 const emxArray_real_T *x, int x_m,
                                 emxArray_real_T *b, int b_m, int nrows,
                                 int ncols, int nrhs, boolean_T varargin_1)
{
  double alpha;
  int b_i;
  int b_remainder;
  int boffset;
  int chunk;
  int i;
  int iend;
  int istart;
  int j;
  int k;
  int nthreads;
  int xoffset;
  if (varargin_1) {
    xoffset = omp_get_num_threads();
    nthreads = (int)floor((double)b_m / (double)ncols);
    if (xoffset < nthreads) {
      nthreads = xoffset;
    }
    xoffset = omp_get_thread_num();
    boffset = xoffset * ncols;
    if (nthreads == 1) {
      istart = 0;
      iend = nrows;
    } else {
      iend = omp_get_thread_num();
      if (iend >= nthreads) {
        istart = 0;
        iend = 0;
      } else {
        chunk = nrows / nthreads;
        b_remainder = nrows - nthreads * chunk;
        if (b_remainder < iend) {
          xoffset = b_remainder;
        } else {
          xoffset = iend;
        }
        istart = iend * chunk + xoffset;
        iend = (istart + chunk) + (iend < b_remainder);
      }
    }
    istart++;
  } else {
    nthreads = 1;
    boffset = 0;
    istart = 1;
    iend = nrows;
  }
  if (istart <= iend) {
    xoffset = -1;
    for (k = 0; k < nrhs; k++) {
      chunk = boffset + 1;
      i = boffset + ncols;
      for (j = chunk; j <= i; j++) {
        b->data[j - 1] = 0.0;
      }
      for (b_i = istart; b_i <= iend; b_i++) {
        alpha = x->data[b_i + xoffset];
        chunk = row_ptr->data[b_i - 1];
        i = row_ptr->data[b_i] - 1;
        for (j = chunk; j <= i; j++) {
          b_remainder = (boffset + col_ind->data[j - 1]) - 1;
          b->data[b_remainder] += alpha * val->data[j - 1];
        }
      }
      xoffset += x_m;
      boffset += b_m;
    }
  }
  if (nthreads > 1) {
#pragma omp barrier
    xoffset = omp_get_num_threads();
    if (xoffset == 1) {
      istart = 0;
      iend = ncols;
    } else {
      iend = omp_get_thread_num();
      chunk = ncols / xoffset;
      b_remainder = ncols - xoffset * chunk;
      if (b_remainder < iend) {
        xoffset = b_remainder;
      } else {
        xoffset = iend;
      }
      istart = iend * chunk + xoffset;
      iend = (istart + chunk) + (iend < b_remainder);
    }
    xoffset = ncols;
    for (j = 2; j <= nthreads; j++) {
      boffset = 1;
      for (k = 0; k < nrhs; k++) {
        chunk = boffset + istart;
        i = (boffset + iend) - 1;
        for (b_i = chunk; b_i <= i; b_i++) {
          b->data[b_i - 1] += b->data[(xoffset + b_i) - 1];
        }
        boffset += b_m;
      }
      xoffset += ncols;
    }
  }
}

static void b_crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                                const emxArray_int32_T *col_ind,
                                const emxArray_real_T *val,
                                const emxArray_real_T *x, int x_m,
                                emxArray_real_T *b, int b_m, int nrows,
                                int nrhs, boolean_T varargin_1)
{
  double t;
  int b_i;
  int b_remainder;
  int chunk;
  int i;
  int iend;
  int istart;
  int j;
  int k;
  int threadID;
  if (varargin_1) {
    iend = omp_get_num_threads();
    if (iend == 1) {
      istart = 0;
      iend = nrows;
    } else {
      threadID = omp_get_thread_num();
      chunk = nrows / iend;
      b_remainder = nrows - iend * chunk;
      if (b_remainder < threadID) {
        iend = b_remainder;
      } else {
        iend = threadID;
      }
      istart = threadID * chunk + iend;
      iend = (istart + chunk) + (threadID < b_remainder);
    }
  } else {
    istart = 0;
    iend = nrows;
  }
  b_remainder = -1;
  threadID = -1;
  for (k = 0; k < nrhs; k++) {
    for (i = istart + 1; i <= iend; i++) {
      t = 0.0;
      chunk = row_ptr->data[i - 1];
      b_i = row_ptr->data[i] - 1;
      for (j = chunk; j <= b_i; j++) {
        t += val->data[j - 1] * x->data[b_remainder + col_ind->data[j - 1]];
      }
      b->data[threadID + i] = t;
    }
    b_remainder += x_m;
    threadID += b_m;
  }
}

static void b_m2c_error(void)
{
  M2C_error("prodAAtx:IncorrectBuffer", "Buffer Ax has incorrect size.");
}

static void c_crs_prodAtAx(const emxArray_int32_T *A_row_ptr,
                           const emxArray_int32_T *A_col_ind,
                           const emxArray_real_T *A_val, int A_nrows,
                           int A_ncols, const emxArray_real_T *x,
                           emxArray_real_T *b)
{
  emxArray_real_T *Ax;
  int i;
  int loop_ub;
  if ((b->size[0] < A_ncols) || (b->size[1] != x->size[1])) {
#pragma omp master
    {
      m2c_error();
    }

  }
  emxInit_real_T(&Ax, 2);
  i = Ax->size[0] * Ax->size[1];
  Ax->size[0] = A_nrows;
  Ax->size[1] = x->size[1];
  emxEnsureCapacity_real_T(Ax, i);
  loop_ub = A_nrows * x->size[1];
  for (i = 0; i < loop_ub; i++) {
    Ax->data[i] = 0.0;
  }
  crs_prodAx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], Ax, A_nrows,
                    A_nrows, x->size[1]);
  b_crs_prodAtx(A_row_ptr, A_col_ind, A_val, A_nrows, A_ncols, Ax, b);
  emxFree_real_T(&Ax);
}

static void c_m2c_error(void)
{
  M2C_error("crs_prodAx:BufferTooSmal",
            "Buffer space for output b is too small.");
}

static void crs_prodAtx(const emxArray_int32_T *A_row_ptr,
                        const emxArray_int32_T *A_col_ind,
                        const emxArray_real_T *A_val, int A_nrows, int A_ncols,
                        const emxArray_real_T *x, emxArray_real_T *b)
{
  int n;
  if ((b->size[0] < A_ncols) || (b->size[1] < x->size[1])) {
    d_m2c_error();
  }
  n = omp_get_num_threads();
  if (n == 1) {
    crs_prodAtx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], b,
                       b->size[0], A_nrows, A_ncols, x->size[1]);
  } else {
    b_crs_prodAtx_kernel(A_row_ptr, A_col_ind, A_val, x, x->size[0], b,
                         b->size[0], A_nrows, A_ncols, x->size[1], true);
  }
}

static void crs_prodAtx_kernel(const emxArray_int32_T *row_ptr,
                               const emxArray_int32_T *col_ind,
                               const emxArray_real_T *val,
                               const emxArray_real_T *x, int x_m,
                               emxArray_real_T *b, int b_m, int nrows,
                               int ncols, int nrhs)
{
  double alpha;
  int b_i;
  int boffset;
  int i;
  int i1;
  int j;
  int k;
  int r;
  int xoffset;
  boffset = 0;
  if (1 <= nrows) {
    xoffset = 0;
    for (k = 0; k < nrhs; k++) {
      i = boffset + 1;
      i1 = boffset + ncols;
      for (j = i; j <= i1; j++) {
        b->data[j - 1] = 0.0;
      }
      for (b_i = 0; b_i < nrows; b_i++) {
        alpha = x->data[b_i + xoffset];
        i = row_ptr->data[b_i];
        i1 = row_ptr->data[b_i + 1] - 1;
        for (j = i; j <= i1; j++) {
          r = (boffset + col_ind->data[j - 1]) - 1;
          b->data[r] += alpha * val->data[j - 1];
        }
      }
      xoffset += x_m;
      boffset += b_m;
    }
  }
}

static void crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                              const emxArray_int32_T *col_ind,
                              const emxArray_real_T *val,
                              const emxArray_real_T *x, int x_m,
                              emxArray_real_T *b, int b_m, int nrows, int nrhs)
{
  double t;
  int b_i;
  int boffset;
  int i;
  int i1;
  int j;
  int k;
  int xoffset;
  xoffset = -1;
  boffset = 0;
  for (k = 0; k < nrhs; k++) {
    for (i = 0; i < nrows; i++) {
      t = 0.0;
      b_i = row_ptr->data[i];
      i1 = row_ptr->data[i + 1] - 1;
      for (j = b_i; j <= i1; j++) {
        t += val->data[j - 1] * x->data[xoffset + col_ind->data[j - 1]];
      }
      b->data[boffset + i] = t;
    }
    xoffset += x_m;
    boffset += b_m;
  }
}

static void d_m2c_error(void)
{
  M2C_error("crs_prodAtx:BufferTooSmal",
            "Buffer space for output b is too small.");
}

static void m2c_error(void)
{
  M2C_error("crs_prodAAtx:IncorrectBuffer", "Buffer b has incorrect size.");
}

static void m2c_warn(void)
{
  M2C_warn("crs_prodAtAx:NestedParallel",
           "You are trying to use nested parallel regions, but nested "
           "parallelism is not enabled.");
}

void crs_prodAtAx(const CRS_Matrix *A, const emxArray_real_T *x,
                  emxArray_real_T *b, emxArray_real_T *Ax,
                  const emxArray_int32_T *nthreads)
{
  int n;
  if ((b->size[0] < A->ncols) || (b->size[1] != x->size[1])) {
#pragma omp master
    {
      m2c_error();
    }

  }
  if ((Ax->size[0] < A->nrows) || (Ax->size[1] != x->size[1])) {
#pragma omp master
    {
      b_m2c_error();
    }

  }
  if (nthreads->size[0] != 0) {
    n = omp_get_nested();
    if (n == 0) {
      n = omp_get_num_threads();
      if ((n > 1) && (nthreads->data[0] > 1)) {
#pragma omp master
        {
          m2c_warn();
        }

      }
    }
#pragma omp parallel default(shared) num_threads(nthreads->data[0])
    {
      if ((Ax->size[0] < A->nrows) || (Ax->size[1] < x->size[1])) {
        c_m2c_error();
      }
      n = omp_get_num_threads();
      if (n == 1) {
        crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], Ax,
                          Ax->size[0], A->nrows, x->size[1]);
      } else {
        b_crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], Ax,
                            Ax->size[0], A->nrows, x->size[1], true);
      }
#pragma omp barrier
      crs_prodAtx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, Ax, b);
    }

  } else {
    n = omp_get_num_threads();
    if (n > 1) {
      if ((Ax->size[0] < A->nrows) || (Ax->size[1] < x->size[1])) {
        c_m2c_error();
      }
      n = omp_get_num_threads();
      if (n == 1) {
        crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], Ax,
                          Ax->size[0], A->nrows, x->size[1]);
      } else {
        b_crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], Ax,
                            Ax->size[0], A->nrows, x->size[1], true);
      }
#pragma omp barrier
      crs_prodAtx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, Ax, b);
    } else {
      if ((Ax->size[0] < A->nrows) || (Ax->size[1] < x->size[1])) {
        c_m2c_error();
      }
      crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], Ax,
                        Ax->size[0], A->nrows, x->size[1]);
      b_crs_prodAtx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, Ax, b);
    }
  }
}

void crs_prodAtAx_initialize(void)
{
}

void crs_prodAtAx_ser(const CRS_Matrix *A, const emxArray_real_T *x,
                      emxArray_real_T *b)
{
  b_crs_prodAtAx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, x, b);
}

void crs_prodAtAx_ser1(const CRS_Matrix *A, const emxArray_real_T *x,
                       emxArray_real_T *b)
{
  c_crs_prodAtAx(A->row_ptr, A->col_ind, A->val, A->nrows, A->ncols, x, b);
}

void crs_prodAtAx_terminate(void)
{
}

File: api/crs/crs_prodAtAx.m
function [b, Ax] = crs_prodAtAx(A, x, b, Ax, nthreads, varargin)
%crs_prodAtAx Compute A'*A*x for a sparse matrix A in CRS format.
%
%      b = crs_prodAtAx(A, x [, b]) or [b, Ax] = crs_prodAtAx(A, x, b, Ax)
% computes in serial. Ax is a buffer for storing the
% intermediate result Ax. If passed as input, b and Ax are assumed
% to have been preallocated.
%
%      [b, Ax] = crs_prodAtAx(A, x, b, Ax, nthreads)
% Computes using nthreads OpenMP threads, where nthreads is an int32.
% Note that size(b,1) should be >= A.ncols*nthreads to facilitate concurrency.
% Otherwise, computation of At*Ax will use up to size(b,1)/A.ncols threads.
%
%      [b, Ax] = crs_prodAtAx(A, x, b, Ax, [])
% computes locally within each OpenMP thread, assuming the
% parallel region has already been initialized. In this mode, Ax must
% be a buffer shared by all threads in the team, and it must be both
% input and output arguments (i.e., passed by reference). Vector
% b must be preallocated, the computation of At*Ax will use up to
% size(b,1)/A.ncols threads.
%
%      [b, Ax] = crs_prodAtAx(A, x, b, Ax, nthreads, comm)
% assumes that A is partitioned rowwise, and x is dupliplicate on all
% processes. It performs a reduction on b within the MPI communicator.
%
%      [b, Ax] = crs_prodAtAx(A, x, b, Ax, nthreads, comm, pgmsg [, pgsz])
% specifies an additiona message that should be summed during
% allreduce and strored in b. This is to piggy-back on allreduce
% to help aggregate small messages to optimize MPI communication.
%
% In multithread or multiprocess mode, this is a collective function,
% and there is an implicit barrier at the end of this function.
%
% The compiled version reqiures A and x to be double precision.
% The M file works for both single and double precision.

%#codegen -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]), 
%#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0])}
%#codegen crs_prodAtAx_ser -args {crs_matrix, coder.typeof(0, [inf,inf])}
%#codegen crs_prodAtAx_ser1 -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}
% %#codegen crs_prodAtAx_mpi -args {crs_matrix, coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), MPI_Comm}
% %#codegen crs_prodAtAx_mpip -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), MPI_Comm, coder.typeof(0, [inf,1])}
% %#codegen crs_prodAtAx_mpip1 -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),
% %#codegen coder.typeof(0, [inf,inf]), coder.typeof(int32(1), [1,1], [1,0]), MPI_Comm, coder.typeof(0, [inf,1]), int32(0)}

coder.inline('never');

assert(nargin<=8);

if nargin<3
    b = zeros(A.ncols,size(x,2)); 
elseif size(b,1)<A.ncols || size(b,2)~=size(x,2)
    OMP_begin_master
    m2c_error('crs_prodAAtx:IncorrectBuffer', 'Buffer b has incorrect size.');
    OMP_end_master
end
if nargin<4
    Ax = zeros(A.nrows,size(x,2));
elseif size(Ax,1)<A.nrows || size(Ax,2)~=size(x,2)
    OMP_begin_master
    m2c_error('prodAAtx:IncorrectBuffer', 'Buffer Ax has incorrect size.');
    OMP_end_master
end

%% Declare parallel region
if nargin>=5 && ~isempty(nthreads)
    if ~ompGetNested && ompGetNumThreads>1 && nthreads(1)>1
        OMP_begin_master
        m2c_warn('crs_prodAtAx:NestedParallel', ...
            'You are trying to use nested parallel regions, but nested parallelism is not enabled.');
        OMP_end_master
    end
    
    OMP_begin_parallel(nthreads(1));
    
    % Computes Ax=A*x
    Ax = crs_prodAx(A, x, Ax, []);
    OMP_barrier;
    
    % Computes b=A'*Ax
    b = crs_prodAtx(A, Ax, b, [], varargin{:});
    
    OMP_end_parallel(Ax);
elseif nargin>=5 && ompGetNumThreads>1
    if nargout<2
        OMP_begin_master
        m2c_warn('crs_prodAtAx:MissingBuffer', ...
            'crs_prodAtAx is called within a parallel region but Ax is not an in+out argument.');
        OMP_end_master
    end
    
    % Computes Ax=A*x
    Ax = crs_prodAx(A, x, Ax, []);
    OMP_barrier;
    
    % Computes b=A'*Ax
    b = crs_prodAtx(A, Ax, b, [], varargin{:});
else
    % Computes Ax=A*x
    Ax = crs_prodAx(A, x, Ax);
    % Computes b=A'*Ax
    b = crs_prodAtx(A, Ax, b);
end

function test %#ok<DEFNU>
%!test
%! if ~exist(['crs_prodAtAx.' mexext], 'file')
%!    m=100; n = 20;
%! else
%!    m=10000; n = 2000;
%! end
%! tic; sp = sprand(m,n,0.5); x=rand(size(sp,2),2);
%! fprintf(1, '\n\tGenerated random matrix in %g seconds\n', toc);
%! tic; b0 = sp'*(sp*x);
%! fprintf(1, '\tComputed reference solution in %g seconds\n', toc);
%! tic; A = crs_matrix(sp);
%! fprintf(1, '\tConverted into crs_matrix in %g seconds\n', toc);

%! fprintf(1, '\tTesting serial: ');
%! tic; b1 = crs_prodAtAx(A, x);
%! fprintf(1, 'Done in %g seconds\n ', toc);
%! assert(norm(b0-b1)<=1.e-10);

%! Ax = zeros(size(sp,1),2);
%! for nthreads=int32([1 2 4 8])
%!     if nthreads>ompGetMaxThreads; break; end
%!     fprintf(1, '\tTesting %d thread(s): ', nthreads);
%!     b2 = zeros(size(sp,2)*nthreads,2);
%!     tic; [b2, Ax] = crs_prodAtAx(A, x, b2, Ax, nthreads);
%!     fprintf(1, 'Done in %g seconds\n ', toc);
%!     assert(norm(b0-b2(1:size(sp,2),:))/norm(b0)<=1.e-10);
%! end

% %! nprocs = double(comm_size(MPI_COMM_WORLD));
% %! fprintf(1, '\tTesting 2 threads with MPI call: ');
% %! b2 = zeros(size(sp,2)*2,2);
% %! tic; [b2, Ax] = crs_prodAtAx(A, x, b2, Ax, int32(2), MPI_COMM_WORLD);
% %! fprintf(1, 'Done in %g seconds\n ', toc);
% %! assert(norm(b0-b2(1:size(sp,2),:))/norm(b0)<=1.e-10);
% 
% %! b3 = zeros(2*size(sp,2)+1,1); Ax = zeros(size(sp,1),1);
% %! fprintf(1, '\tTesting 2 threads with piggy-back: ');
% %! tic; [b3, Ax] = crs_prodAtAx(A, x(:,1), b3, Ax, int32(2), MPI_COMM_WORLD, 1, int32(1));
% %! fprintf(1, 'Done in %g seconds\n ', toc);
% %! assert(nprocs>1 || norm(b0(:,1)-b3(1:size(sp,2)))/norm(b0)<=1.e-10);
% %! assert(b3(size(sp,2)+1)==nprocs);
